pkg/status/versionchecks.go

```diff
@@ -0,0 +1,60 @@
+package status
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"reflect"
+	"time"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	"github.com/google/go-github/v52/github"
+	errs "github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// ErrMsgDeploymentIsNotUpToDate means that deployment version is not aligned with source code version
+	ErrMsgDeploymentIsNotUpToDate = "deployment version is not up to date with latest github commit SHA"
+
+	// DeploymentThreshold is the threshold after which we can be almost sure the deployment was not updated on the cluster with the latest version/commit,
+	// in this case some issue is preventing the new deployment to happen.
+	DeploymentThreshold = 30 * time.Minute
+)
+
+// CheckDeployedVersionIsUpToDate verifies if there is a match between the latest commit in Github for a given repo and branch matches the provided commit SHA.
+// There is some preconfigured delay/threshold that we keep in account before returning an `error condition`.
+func CheckDeployedVersionIsUpToDate(githubClient *github.Client, repoName, repoBranch, deployedCommitSHA string) *toolchainv1alpha1.Condition {
+	// get the latest commit from given repository and branch
```

Comment:
I have one proposal how we could make the code on the controller's side easier. When I mentioned in one of my comments that we could move entirely the check for the last call to toolchain-common, I mean everything. Looking at the code, we could move even more and simplify the code duplication on the controller side:
```suggestion

type VersionCheckManager struct {
	GetGithubClientFunc client.GetGitHubClientFunc
	lastGHCallsPerRepo  map[string]time.Time
}

// CheckDeployedVersionIsUpToDate verifies if there is a match between the latest commit in Github for a given repo and branch matches the provided commit SHA.
// There is some preconfigured delay/threshold that we keep in account before returning an `error condition`.
func (m *VersionCheckManager) CheckDeployedVersionIsUpToDate(isProd bool, accessTokenKey string, alreadyExistingConditions []toolchainv1alpha1.Condition, repoName, repoBranch, deployedCommitSHA string) *toolchainv1alpha1.Condition {
	// the first two checks are pretty much the same for all components
	if !isProd {
		cond := NewDeploymentVersionUpToDateCondition(toolchainv1alpha1.ToolchainStatusDeploymentVersionCheckDisabledReason)
		// you could change the NewDeploymentVersionUpToDateCondition function (or create a new one) so it would accept also the message
		cond.Message = "is not running in prod environment"
		return cond
	}
	if accessTokenKey == "" {
		cond := NewDeploymentVersionUpToDateCondition(toolchainv1alpha1.ToolchainStatusDeploymentVersionCheckDisabledReason)
		cond.Message = "access token key is not provided"
		return cond
	}
	// we can store the last call per repo name, so it will solve the gaps between calls for host & reg-service which is done form the same controller
	if m.lastGHCallsPerRepo == nil {
		m.lastGHCallsPerRepo = map[string]time.Time{}
	}
	lastCall, present := m.lastGHCallsPerRepo[repoName]
	if present && !client.CanIssueGitHubRequest(lastCall) {
		// we need to make the gap between the calls for this repo, but we have to set the condition to something - let's reuse what was set last time
		previouslySet, found := condition.FindConditionByType(alreadyExistingConditions, toolchainv1alpha1.ToolchainStatusDeploymentVersionCheckDisabledReason)
		if !found {
			cond := NewDeploymentVersionUpToDateCondition("unknown")
			cond.Message = "waiting for the next attempt"
			return cond
		}
		return &previouslySet
	}
	m.lastGHCallsPerRepo[repoName] = time.Now()

	githubClient := m.GetGithubClientFunc(accessTokenKey)
	// get the latest commit from given repository and branch
```

on the controller's side, you would initiate the manager as part of creating the Reconciler:
```go
if err := (&toolchainstatus.Reconciler{
	Client:         mgr.GetClient(),
	Scheme:         mgr.GetScheme(),
	HTTPClientImpl: &http.Client{},
	VersionCheckManager: status.VersionCheckManager{
		GetGithubClientFunc: commonclient.NewGitHubClient,
	},
	GetMembersFunc: commoncluster.GetMemberClusters,
	Namespace:      namespace,
}).SetupWithManager(mgr); err != nil {
	setupLog.Error(err, "unable to create controller", "controller", "ToolchainStatus")
	os.Exit(1)
}
```
so it could be easily overridden in unit tests